tests/main/snapd-snap/task.yaml

```diff
@@ -1,27 +1,144 @@
     snap install snapcraft --candidate --classic
+    tests.cleanup defer snap remove snapcraft
+
+    if [ "$SNAPCRAFT_BUILD_ENVIRONMENT" = "lxd" ]; then
+        echo "Remove any installed debs (some images carry them) to ensure we test the snap"
+        # apt -v to test if apt is usable
+        if command -v apt && apt -v; then
+            # meh trusty's apt doesn't support -y, so use apt-get
+            apt-get autoremove -y lxd
+            if ! os.query is-debian-sid; then
+                # no lxd-client on debian sid
+                apt-get autoremove -y lxd-client
+            fi
+        fi
+
+        # load the fuse kernel module before installing lxd
+        modprobe fuse
+
+        echo "Install lxd"
+        snap install lxd --channel="$LXD_SNAP_CHANNEL"
+        tests.cleanup defer snap remove lxd
+
+        echo "Setup the lxd snap"
+        snap set lxd waitready.timeout=240
+        lxd waitready
+        lxd init --auto
+
+        echo "Setting up proxy for lxc"
+        if [ -n "${http_proxy:-}" ]; then
+            lxd.lxc config set core.proxy_http "$http_proxy"
+        fi
+        if [ -n "${https_proxy:-}" ]; then
+            lxd.lxc config set core.proxy_https "$http_proxy"
+        fi
+
+        echo "Switch the core snap back due to spread system prepare shenanigans"
+        # When we run spread, we will repack the core snap that is installed on
+        # the system to match the binaries that we built on this system so that
+        # we keep executing the stuff from the branch, even if re-exec is 
+        # enabled unconditionally for all spread tests. This presents a problem
+        # for snapcraft, which downloads the current snap file from snapd 
+        # directly using the /v2/snaps/{name}/file endpoint, which will just 
+        # send over the modified file without indicating at all that the current
+        # file is modified from the installed revision (indeed snapd reports to
+        # snapcraft that it is running the specific edge revision from the 
+        # store, and not x1 as it would in devmode for example which snapcraft
+        # correctly handles).
+        # As such, we need to switch the modified core snap on the system back 
+        # with the original core snap on the system that is unmodified and will
+        # match the expected assertions that snapcraft calculates
+
+        # the easy way is to trick snapd is to just mount back the original file
+        # on top of the repacked one, then snapd will just serve the correct
+        # file over HTTP to snapcraft when it asks for it
+
+        CORE_SNAP_REV=$(snap list core | grep -v Name | awk '{print $3}')
+        CORE_SNAP="/var/lib/snapd/snaps/core_${CORE_SNAP_REV}.snap"
+        ORIG_CORE_SNAP=${CORE_SNAP}.orig
+        
+        mount --bind "$ORIG_CORE_SNAP" "$CORE_SNAP"
+        tests.cleanup defer umount "$CORE_SNAP"
+    fi
 
 restore: |
+    # shellcheck disable=SC2164
     cd "$PROJECT_PATH"
```

Comment:
do we need this?

Reply:
sort of, we need to cleanup after snapcraft via `snapcraft clean`, which needs to be run from $PROJECT_PATH, but `tests.cleanup defer sh -c 'cd $PROJECT_PATH && snapcraft clean'` ends up being executed as two separate statements somehow with snapcraft clean being just run in $PWD and not in $PROJECT_PATH. I have a fix for this to just put these two things into a script and have defer execute that script instead, but it would be great if you could work on adding a --cwd argument to tests.cleanup to support changing directories when a command is being executed, that would be super helpful for instances like this